From 98e0d9fe0b2f8dd007d71e3649f8eba9ec6eb4eb Mon Sep 17 00:00:00 2001
From: Sushil Sati <v-sushilsati@microsoft.com>
Date: Thu, 17 Sep 2026 12:01:03 +0000
Subject: [PATCH] Add Go toolchain compatibility fixes

Newer Go toolchains reject references to the private runtime.roundupsize
symbol. Copy the allocator size-class calculation used by newer Vitess
releases instead of linking to the runtime implementation. Also avoid
rendering Go-version-dependent sync.Mutex internals in schemaz output and
use explicit formats for dynamic error strings required by newer Go vet.

Upstream reference for the roundupsize/size-class implementation:
https://github.com/vitessio/vitess/blob/main/go/hack/msize.go
---
 go/hack/atof_go127.go                         | 11 +++++
 go/hack/atof_legacy.go                        | 11 +++++
 go/hack/runtime.go                            | 44 +++++++++++++++----
 go/mysql/fastparse/fastparse.go               | 11 +++++
 go/mysql/sqlerror/sql_error.go                |  2 +-
 go/vt/mysqlctl/s3backupstorage/s3.go          |  2 +-
 go/vt/vttablet/tabletserver/schema/schema.go  |  7 ++-
 .../tabletserver/schema/schemaz_test.go       |  2 +-
 .../tabletserver/stateful_connection.go       |  2 +-
 go/vt/wrangler/keyspace.go                    |  2 +-
 go/vt/wrangler/materializer.go                |  2 +-
 go/vt/wrangler/traffic_switcher.go            | 10 ++---
 12 files changed, 83 insertions(+), 23 deletions(-)
 create mode 100644 go/hack/atof_go127.go
 create mode 100644 go/hack/atof_legacy.go

diff --git a/go/hack/atof_go127.go b/go/hack/atof_go127.go
new file mode 100644
index 0000000..f73b277
--- /dev/null
+++ b/go/hack/atof_go127.go
@@ -0,0 +1,11 @@
+//go:build go1.27
+
+package hack
+
+import _ "unsafe"
+
+//go:linkname Atof64 internal/strconv.atof64
+func Atof64(s string) (float64, int, error)
+
+//go:linkname Atof32 internal/strconv.atof32
+func Atof32(s string) (float32, int, error)
diff --git a/go/hack/atof_legacy.go b/go/hack/atof_legacy.go
new file mode 100644
index 0000000..6ecb63c
--- /dev/null
+++ b/go/hack/atof_legacy.go
@@ -0,0 +1,11 @@
+//go:build !go1.27
+
+package hack
+
+import _ "unsafe"
+
+//go:linkname Atof64 strconv.atof64
+func Atof64(s string) (float64, int, error)
+
+//go:linkname Atof32 strconv.atof32
+func Atof32(s string) (float32, int, error)
diff --git a/go/hack/runtime.go b/go/hack/runtime.go
index 8342850..8e3d484 100644
--- a/go/hack/runtime.go
+++ b/go/hack/runtime.go
@@ -22,6 +22,22 @@ import (
 	"unsafe"
 )
 
+const (
+	_MaxSmallSize   = 32768
+	smallSizeDiv    = 8
+	smallSizeMax    = 1024
+	largeSizeDiv    = 128
+	_NumSizeClasses = 68
+	_PageShift      = 13
+	_PageSize       = 1 << _PageShift
+)
+
+var (
+	class_to_size    = [_NumSizeClasses]uint16{0, 8, 16, 24, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240, 256, 288, 320, 352, 384, 416, 448, 480, 512, 576, 640, 704, 768, 896, 1024, 1152, 1280, 1408, 1536, 1792, 2048, 2304, 2688, 3072, 3200, 3456, 4096, 4864, 5376, 6144, 6528, 6784, 6912, 8192, 9472, 9728, 10240, 10880, 12288, 13568, 14336, 16384, 18432, 19072, 20480, 21760, 24576, 27264, 28672, 32768}
+	size_to_class8   = [smallSizeMax/smallSizeDiv + 1]uint8{0, 1, 2, 3, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17, 18, 18, 19, 19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 22, 23, 23, 23, 23, 24, 24, 24, 24, 25, 25, 25, 25, 26, 26, 26, 26, 27, 27, 27, 27, 27, 27, 27, 27, 28, 28, 28, 28, 28, 28, 28, 28, 29, 29, 29, 29, 29, 29, 29, 29, 30, 30, 30, 30, 30, 30, 30, 30, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32}
+	size_to_class128 = [(_MaxSmallSize-smallSizeMax)/largeSizeDiv + 1]uint8{32, 33, 34, 35, 36, 37, 37, 38, 38, 39, 39, 40, 40, 40, 41, 41, 41, 42, 43, 43, 44, 44, 44, 44, 44, 45, 45, 45, 45, 45, 45, 46, 46, 46, 46, 47, 47, 47, 47, 47, 47, 48, 48, 48, 49, 49, 50, 51, 51, 51, 51, 51, 51, 51, 51, 51, 51, 52, 52, 52, 52, 52, 52, 52, 52, 52, 52, 53, 53, 54, 54, 54, 54, 55, 55, 55, 55, 55, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 57, 57, 57, 57, 57, 57, 57, 57, 57, 57, 58, 58, 58, 58, 58, 58, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 61, 61, 61, 61, 61, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 62, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 66, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67, 67}
+)
+
 //go:noescape
 //go:linkname strhash runtime.strhash
 func strhash(p unsafe.Pointer, h uintptr) uintptr
@@ -33,19 +49,31 @@ func RuntimeStrhash(str string, seed uint64) uint64 {
 	return uint64(strhash(unsafe.Pointer(&str), uintptr(seed)))
 }
 
-//go:linkname roundupsize runtime.roundupsize
-func roundupsize(size uintptr) uintptr
+func alignUp(n, a uintptr) uintptr {
+	return (n + a - 1) &^ (a - 1)
+}
+
+func divRoundUp(n, a uintptr) uintptr {
+	return (n + a - 1) / a
+}
+
+func roundupsize(size uintptr) uintptr {
+	if size < _MaxSmallSize {
+		if size <= smallSizeMax-8 {
+			return uintptr(class_to_size[size_to_class8[divRoundUp(size, smallSizeDiv)]])
+		}
+		return uintptr(class_to_size[size_to_class128[divRoundUp(size-smallSizeMax, largeSizeDiv)]])
+	}
+	if size+_PageSize < size {
+		return size
+	}
+	return alignUp(size, _PageSize)
+}
 
 // RuntimeAllocSize returns size of the memory block that mallocgc will allocate if you ask for the size.
 func RuntimeAllocSize(size int64) int64 {
 	return int64(roundupsize(uintptr(size)))
 }
 
-//go:linkname Atof64 strconv.atof64
-func Atof64(s string) (float64, int, error)
-
-//go:linkname Atof32 strconv.atof32
-func Atof32(s string) (float32, int, error)
-
 //go:linkname FastRand runtime.fastrand
 func FastRand() uint32
diff --git a/go/mysql/fastparse/fastparse.go b/go/mysql/fastparse/fastparse.go
index a17401b..1ef8d3b 100644
--- a/go/mysql/fastparse/fastparse.go
+++ b/go/mysql/fastparse/fastparse.go
@@ -232,6 +232,17 @@ func ParseFloat64(s string) (float64, error) {
 	// string as possible. This functionality is implemented in the `strconv` package
 	// of the standard library, but not exposed, so we hook into it.
 	val, l, err := hack.Atof64(s[ws:])
+	if err != nil && !errors.Is(err, strconv.ErrRange) && !errors.Is(err, strconv.ErrSyntax) {
+		kind := strconv.ErrSyntax
+		if math.IsInf(val, 0) {
+			kind = strconv.ErrRange
+		}
+		err = &strconv.NumError{
+			Func: "ParseFloat",
+			Num:  s[ws:],
+			Err:  kind,
+		}
+	}
 	for l < len(s[ws:]) {
 		if !isSpace(s[ws+uint(l)]) {
 			break
diff --git a/go/mysql/sqlerror/sql_error.go b/go/mysql/sqlerror/sql_error.go
index fc201be..9b5de17 100644
--- a/go/mysql/sqlerror/sql_error.go
+++ b/go/mysql/sqlerror/sql_error.go
@@ -282,7 +282,7 @@ func convertToMysqlError(err error) error {
 	if !ok {
 		return err
 	}
-	return NewSQLError(mysqlCode.num, mysqlCode.state, err.Error())
+	return NewSQLError(mysqlCode.num, mysqlCode.state, "%s", err.Error())
 }
 
 var isGRPCOverflowRE = regexp.MustCompile(`.*?grpc: (received|trying to send) message larger than max \(\d+ vs. \d+\)`)
diff --git a/go/vt/mysqlctl/s3backupstorage/s3.go b/go/vt/mysqlctl/s3backupstorage/s3.go
index cdc4e35..4132922 100644
--- a/go/vt/mysqlctl/s3backupstorage/s3.go
+++ b/go/vt/mysqlctl/s3backupstorage/s3.go
@@ -412,7 +412,7 @@ func (bs *S3BackupStorage) RemoveBackup(ctx context.Context, dir, name string) e
 		}
 
 		for _, objError := range out.Errors {
-			return fmt.Errorf(objError.String())
+			return fmt.Errorf("%s", objError.String())
 		}
 
 		if objs.NextContinuationToken == nil {
diff --git a/go/vt/vttablet/tabletserver/schema/schema.go b/go/vt/vttablet/tabletserver/schema/schema.go
index 4b3d9c8..490b3d3 100644
--- a/go/vt/vttablet/tabletserver/schema/schema.go
+++ b/go/vt/vttablet/tabletserver/schema/schema.go
@@ -17,11 +17,10 @@ limitations under the License.
 package schema
 
 import (
+	"fmt"
 	"sync"
 	"time"
 
-	"vitess.io/vitess/go/vt/log"
-
 	"vitess.io/vitess/go/vt/sqlparser"
 
 	querypb "vitess.io/vitess/go/vt/proto/query"
@@ -84,10 +83,10 @@ func (seq *SequenceInfo) Reset() {
 	seq.LastVal = 0
 }
 
-func (seq *SequenceInfo) String() {
+func (seq *SequenceInfo) String() string {
 	seq.Lock()
 	defer seq.Unlock()
-	log.Infof("SequenceInfo: NextVal: %d, LastVal: %d", seq.NextVal, seq.LastVal)
+	return fmt.Sprintf("SequenceInfo: NextVal: %d, LastVal: %d", seq.NextVal, seq.LastVal)
 }
 
 // MessageInfo contains info specific to message tables.
diff --git a/go/vt/vttablet/tabletserver/schema/schemaz_test.go b/go/vt/vttablet/tabletserver/schema/schemaz_test.go
index 0127c79..472b107 100644
--- a/go/vt/vttablet/tabletserver/schema/schemaz_test.go
+++ b/go/vt/vttablet/tabletserver/schema/schemaz_test.go
@@ -50,7 +50,7 @@ func TestSchamazHandler1(t *testing.T) {
 		`<td>id: INT32<br>next_id: INT64<br>cache: INT64<br>increment: INT64<br></td>`,
 		`<td>id<br></td>`,
 		`<td>sequence</td>`,
-		`<td>{{0 0} 0 0}&lt;nil&gt;</td>`,
+		`<td>SequenceInfo: NextVal: 0, LastVal: 0&lt;nil&gt;</td>`,
 	}
 	matched, err = regexp.Match(strings.Join(seq, `\s*`), body)
 	require.NoError(t, err)
diff --git a/go/vt/vttablet/tabletserver/stateful_connection.go b/go/vt/vttablet/tabletserver/stateful_connection.go
index 067f219..24d5bbc 100644
--- a/go/vt/vttablet/tabletserver/stateful_connection.go
+++ b/go/vt/vttablet/tabletserver/stateful_connection.go
@@ -159,7 +159,7 @@ func (sc *StatefulConnection) unlock(updateTime bool) {
 // Release is used when the connection will not be used ever again.
 // The underlying dbConn is removed so that this connection cannot be used by mistake.
 func (sc *StatefulConnection) Release(reason tx.ReleaseReason) {
-	sc.Releasef(reason.String())
+	sc.Releasef("%s", reason.String())
 }
 
 // Releasef is used when the connection will not be used ever again.
diff --git a/go/vt/wrangler/keyspace.go b/go/vt/wrangler/keyspace.go
index a5f7d6a..d0a774a 100644
--- a/go/vt/wrangler/keyspace.go
+++ b/go/vt/wrangler/keyspace.go
@@ -82,7 +82,7 @@ func (wr *Wrangler) validateNewWorkflow(ctx context.Context, keyspace, workflow
 					return
 				}
 				if p3qr != nil && len(p3qr.Rows) != 0 {
-					allErrors.RecordError(vterrors.Wrap(fmt.Errorf(validation.msg), "validateWorkflowName.VReplicationExec"))
+					allErrors.RecordError(vterrors.Wrap(fmt.Errorf("%s", validation.msg), "validateWorkflowName.VReplicationExec"))
 					return
 				}
 			}
diff --git a/go/vt/wrangler/materializer.go b/go/vt/wrangler/materializer.go
index 5b159e3..a22adeb 100644
--- a/go/vt/wrangler/materializer.go
+++ b/go/vt/wrangler/materializer.go
@@ -355,7 +355,7 @@ func (wr *Wrangler) MoveTables(ctx context.Context, workflow, sourceKeyspace, ta
 				migrationID, strings.Join(tablets, ","))
 			msg += fmt.Sprintf("please review and delete it before proceeding and restart the workflow using the Workflow %s.%s start",
 				workflow, targetKeyspace)
-			return fmt.Errorf(msg)
+			return fmt.Errorf("%s", msg)
 		}
 	}
 	if autoStart {
diff --git a/go/vt/wrangler/traffic_switcher.go b/go/vt/wrangler/traffic_switcher.go
index a6b3587..4aaa924 100644
--- a/go/vt/wrangler/traffic_switcher.go
+++ b/go/vt/wrangler/traffic_switcher.go
@@ -335,7 +335,7 @@ func (wr *Wrangler) SwitchReads(ctx context.Context, targetKeyspace, workflowNam
 	cells []string, direction workflow.TrafficSwitchDirection, dryRun bool) (*[]string, error) {
 	// Consistently handle errors by logging and returning them.
 	handleError := func(message string, err error) (*[]string, error) {
-		werr := vterrors.Errorf(vtrpcpb.Code_INTERNAL, fmt.Sprintf("%s: %v", message, err))
+		werr := vterrors.Errorf(vtrpcpb.Code_INTERNAL, "%s: %v", message, err)
 		wr.Logger().Error(werr)
 		return nil, werr
 	}
@@ -346,7 +346,7 @@ func (wr *Wrangler) SwitchReads(ctx context.Context, targetKeyspace, workflowNam
 	}
 	if ts == nil {
 		errorMsg := fmt.Sprintf("workflow %s not found in keyspace %s", workflowName, targetKeyspace)
-		return handleError("failed to get the current state of the workflow", fmt.Errorf(errorMsg))
+		return handleError("failed to get the current state of the workflow", fmt.Errorf("%s", errorMsg))
 	}
 	log.Infof("Switching reads: %s.%s tt %+v, cells %+v, workflow state: %+v", targetKeyspace, workflowName, servedTypes, cells, ws)
 	var switchReplicas, switchRdonly bool
@@ -478,7 +478,7 @@ func (wr *Wrangler) SwitchWrites(ctx context.Context, targetKeyspace, workflowNa
 	cancel, reverse, reverseReplication bool, dryRun, initializeTargetSequences bool) (journalID int64, dryRunResults *[]string, err error) {
 	// Consistently handle errors by logging and returning them.
 	handleError := func(message string, err error) (int64, *[]string, error) {
-		werr := vterrors.Errorf(vtrpcpb.Code_INTERNAL, fmt.Sprintf("%s: %v", message, err))
+		werr := vterrors.Errorf(vtrpcpb.Code_INTERNAL, "%s: %v", message, err)
 		wr.Logger().Error(werr)
 		return 0, nil, werr
 	}
@@ -490,7 +490,7 @@ func (wr *Wrangler) SwitchWrites(ctx context.Context, targetKeyspace, workflowNa
 	}
 	if ts == nil {
 		errorMsg := fmt.Sprintf("workflow %s not found in keyspace %s", workflowName, targetKeyspace)
-		return handleError("failed to get the current workflow state", fmt.Errorf(errorMsg))
+		return handleError("failed to get the current workflow state", fmt.Errorf("%s", errorMsg))
 	}
 
 	var sw iswitcher
@@ -1723,7 +1723,7 @@ func doValidateWorkflowHasCompleted(ctx context.Context, ts *trafficSwitcher) er
 		_ = ts.ForAllSources(func(source *workflow.MigrationSource) error {
 			wg.Add(1)
 			if source.GetShard().IsPrimaryServing {
-				rec.RecordError(fmt.Errorf(fmt.Sprintf("Shard %s is still serving", source.GetShard().ShardName())))
+				rec.RecordError(fmt.Errorf("Shard %s is still serving", source.GetShard().ShardName()))
 			}
 			wg.Done()
 			return nil
-- 
2.45.4

